docs/backend/behaviors.md

```diff
@@ -11,12 +11,320 @@ myst:
+
+### Creating a behavior with an adapter and factory
+
+Given you want to display a price with different content types.
+The price is stored as the net value on the type as a floating point number.
+For display, you need at several places the value added tax (VAT) and the gross value.
+
+You create a schema with the net value for the form and attributes for the calculated values.
+You create an adapter to calculate the VAT and gross values.
+You need a marker interface to distinguish between context and adapter.
+
+Add Python code in the file `price.py`:
+
+```python
+from plone.autoform.interfaces import IFormFieldProvider
+from plone.supermodel import model
+from zope import schema
+from zope.interface import Attribute
+from zope.interface import implementer
+from zope.interface import Interface
+from zope.interface import provider
+
+
+@provider(IFormFieldProvider)
+class IPriceBehavior(model.Schema):
+    """Behavior: a price, VAT and gross."""
+
+    price_net = schema.Float(
+        title="Price (net)",
+        required=True,
+    )
+    price_vat = Attribute("VAT 20% of net price")
+    price_gross = Attribute("Price gross (net + VAT 20%")
+
+
+class IPriceMarker(Interface):
+    """Marker for content that has a price."""
+
+
+@implementer(IPriceBehavior)
+class PriceAdapter:
+    def __init__(self, context):
+        self.context = context
+
+    @property
+    def price_net(self):
+        """Getter, read from context and return back"""
+        return self.context.price_net
+
+    @price_net.setter
+    def price_net(self, value):
+        """Setter, called by the form, set on context"""
+        self.context.price_net = value
+
+    @property
+    def price_vat(self):
+        return self.price_net * 0.2
+
+    @property
+    def price_gross(self):
+        return self.price_net + self.price_vat
+
+```
+
+The registration in the `configure.zcml` looks like so:
+
+```XML
+  <plone:behavior
+      factory=".price.PriceAdapter"
+      for=".price.IPriceMarker"
+      marker=".price.IPriceMarker"
+      name="myproject.price"
+      provides=".price.IPriceBehavior"
+      title="Price with net, VAT and gross"
+  />
+  ```
+
+After a restart of Plone, the behavior can be added to the content type in the {guilabel}`Content Types` control panel.
+The add and edit forms contain a new field `Price (net)`.
+
+This field is not displayed in most views.
+To display the data entered in this field, you need to modify the page template by adding the `price_net` field as `context.price_net`.
+To access the `price_vat` and `price_gross` fields, you need to get the adapter in your view class:
+
+```Python
+from .price import IPriceBehavior
+
+class SomeViewClass:
+
+    def vat(self):
+        adapter = IPriceBehavior(context)
+        return adapter.price_vat
+
+    def gross(self):
+        adapter = IPriceBehavior(context)
+        return adapter.price_gross
+```
```

Comment:
```suggestion
```Python
from .price import IPriceBehavior

class SomeViewClass:

    def vat(self):
        adapted_context = IPriceBehavior(context)
        return adapted_context.price_vat

    def gross(self):
        adapted_context = IPriceBehavior(context)
        return adapted_context.price_gross
```
```

Reply:
OK, with this code i see what you want to express. but its confusing, at least to me.

Reply:
any other opinions?

Reply:
Please lowercase `Python` for the lexer.

I don't have an opinion about the code itself. I don't understand what you find confusing about @ksuess's suggestion.

In my brain (which is a scary place for anyone to dwell), using `adapter` implies that the thing can do adaptations, whereas `adapted_context` implies the context has been adapted and leaving it indeterminate whether it can do adaptations.

To be honest, this is all confusing, so maybe it is less/as/more confusing with the suggestion? I don't know.

Reply:
OK, what about just `adaptet`? In my brain, not much different from your, the idea pops up I get a context, which is not the case.

Reply:
> Please lowercase Python for the lexer.

👍🏼

Reply:
> OK, what about just `adapted`? In my brain, not much different from yours, the idea pops up I get a context, which is not the case.

Hm, I see your point. `adapted` might be part of the name of the thing. Maybe the answer to this question will yield the correct answer?

"When you pass a context to a behavior interface, what thing do you get in return?"

Reply:
so what happens there?

1. An abstract factory is called. (IPriceBehavior) Yes its an interface, but here in its shortcut as a factory usage, to do an getAdapter(context, IPriceBehavior). We could call latter with the exact same result. 
2. The factored adapter specific for the given content item is assigned to variable. 

So its an adapter, like when you travel the world and you have a German Schuko plug in youd notebooks power cable. I  carry a bag with adapter for all the different power outlets with me. Now, on my stop in Italy I find a Type L outlet. I say getAdapter by looking at the outlet (ITypeL) , what my plug needs (ISchuko) and then choose the right one. Then I plug the adapter in the wall. This is what the abstract factory does and return. Now I use the new interface. It provides (ISchuko) and plug in my Schuko plug ti get power.

Reply:
Question here is, what does the call return? `IPriceBehavior(context)` which is similar to  `getAdapter(context, IPriceBehavior)`. It returns an adapter bound to the context. So if we stick to Zope component architectures naming (which IMO is a good idea) this is the actual instance of the "adapter". It does not return something adapted, which implies we still have this something there, but somehow enhanced, like wrapped. But we do not have a wrapper here (also a design pattern) but an "object adapter pattern". So we get something new, nothing polymorphic.
`adapted_context` sounds to my brain like its context, but provides something _addtional_ - which is not the case. It only provides the provided interface, nothing else.
We could call it:
- as is `adapter` - because that's what its is in ZCA
- more specific `behavior_adapter` - which indicates its the behavior specific adapter we get
- `price_for_context` the semantic way to tell what we get (I like this one and would use it in my code) or 
- more general `price_adapter`.

Reply:
> `price_for_context` the semantic way to tell what we get (I like this one and would use it in my code) or

We have a winner! 🔔

Reply:
Done in 00e56f3db0ae85ebe1c4433510787c5545c1a272